Source Language is English and Target language is Hindi

In [1]:
!pip install transformers
!pip install transformers sentencepiece
!pip install transformers datasets
!pip install datasets
!pip install sacrebleu

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 507.1/507.1 kB 8.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 14.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 7.9 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 9.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.3/106.3 kB 4.5 MB/s eta 0:00:00


In [2]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM
import pickle
from transformers import pipeline
import pandas as pd
import csv
from datasets import load_dataset
import matplotlib.pyplot as plt
import seaborn as sns

Loading Flores 200 dataset for English Language

In [3]:
from datasets import load_dataset
import pandas as pd

# Specify the dataset name and configuration
dataset_name = 'Muennighoff/flores200'
config_name = 'eng_Latn'  # Choose a specific configuration from the available list

# Load the dataset with the specified configuration
dataset = load_dataset(dataset_name, config_name)

# Choose the appropriate split key ('dev' or 'devtest')
split_key = 'dev'  # You can change this to 'devtest' if you prefer that split

# Access the selected split
selected_split = dataset[split_key]

# Convert the selected split to a Pandas DataFrame
df = pd.DataFrame(selected_split)

# Print the first 5 rows of the DataFrame
print(df.head())

# Save the DataFrame to a CSV file in the local environment
csv_filename = 'flores200_data_English.csv'
df.to_csv(csv_filename, index=False)
print(f"Data has been saved to {csv_filename}")

# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(csv_filename)

# Print the first 5 rows of the DataFrame to verify the data has been loaded successfully
print(df.head())

# Get the data formats of specific columns
columns_of_interest = ['id', 'URL', 'domain', 'topic', 'has_image', 'has_hyperlink', 'sentence']
data_formats = df[columns_of_interest].dtypes

# Print the data formats of the specified columns
print(data_formats)

# Specify the columns of interest excluding 'URL' and 'sentence'
columns_of_interest = ['id', 'domain', 'topic', 'has_image', 'has_hyperlink']

# Calculate the number of unique values (categories) for each column
unique_value_counts = df[columns_of_interest].nunique()

# Print the number of unique values in each column
print(unique_value_counts)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/datasets/load.py:1429: FutureWarning: The repository for Muennighoff/flores200 contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/Muennighoff/flores200
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`

Generating dev split: 0 examples [00:00, ? examples/s]

Generating devtest split: 0 examples [00:00, ? examples/s]

   id                                                URL    domain     topic  \
0   1  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
1   2  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
2   3  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
3   4  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
4   5  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   

   has_image  has_hyperlink                                           sentence  
0          0              0  On Monday, scientists from the Stanford Univer...  
1          0              0  Lead researchers say this may bring early dete...  
2          0              0  The JAS 39C Gripen crashed onto a runway at ar...  
3          0              0  The pilot was identified as Squadron Leader Di...  
4          0              0  Local media reports an airport fire vehicle ro...  
Data has been saved to flores200_

Loading Flores 200 dataset for Hindi Language


In [4]:
from datasets import load_dataset
import pandas as pd

# Specify the dataset name and configuration
dataset_name = 'Muennighoff/flores200'
config_name = 'hin_Deva'  # Choose a specific configuration from the available list

# Load the dataset with the specified configuration
dataset = load_dataset(dataset_name, config_name)

# Choose the appropriate split key ('dev' or 'devtest')
split_key = 'dev'  # You can change this to 'devtest' if you prefer that split

# Access the selected split
selected_split = dataset[split_key]

# Convert the selected split to a Pandas DataFrame
df = pd.DataFrame(selected_split)

# Print the first 5 rows of the DataFrame
print(df.head())

# Save the DataFrame to a CSV file in the local environment
csv_filename = f'flores200_data_{config_name}.csv'
df.to_csv(csv_filename, index=False)
print(f"Data has been saved to {csv_filename}")

# Read the CSV file into a Pandas DataFrame
df = pd.read_csv(csv_filename)

# Print the first 5 rows of the DataFrame to verify the data has been loaded successfully
print(df.head())

# Get the data formats of specific columns
columns_of_interest = ['id', 'URL', 'domain', 'topic', 'has_image', 'has_hyperlink', 'sentence']
data_formats = df[columns_of_interest].dtypes

# Print the data formats of the specified columns
print(data_formats)

# Specify the columns of interest excluding 'URL' and 'sentence'
columns_of_interest = ['id', 'domain', 'topic', 'has_image', 'has_hyperlink']

# Calculate the number of unique values (categories) for each column
unique_value_counts = df[columns_of_interest].nunique()

# Print the number of unique values in each column
print(unique_value_counts)


/usr/local/lib/python3.10/dist-packages/datasets/load.py:1429: FutureWarning: The repository for Muennighoff/flores200 contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/Muennighoff/flores200
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


Generating dev split: 0 examples [00:00, ? examples/s]

Generating devtest split: 0 examples [00:00, ? examples/s]

   id                                                URL    domain     topic  \
0   1  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
1   2  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
2   3  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
3   4  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
4   5  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   

   has_image  has_hyperlink                                           sentence  
0          0              0  सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ म...  
1          0              0  शोधकर्ताओं ने कहा है कि यह अल्प आय वाले देशों ...  
2          0              0  स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बज...  
3          0              0  पायलट की पहचान स्क्वाड्रन लीडर दिलोकृत पटावी क...  
4          0              0  स्थानीय मीडिया ने बताया है कि कार्रवाई करने के...  
Data has been saved to flores200_

Merging two datasets

In [5]:
# Load English DataFrame
df_english = pd.read_csv('flores200_data_English.csv')

# Load Hindi DataFrame
df_hindi = pd.read_csv('flores200_data_hin_Deva.csv')

# Merge DataFrames based on common columns
merged_df = pd.merge(df_english, df_hindi, on=['id', 'domain', 'topic'])

# Rename 'sentence' columns
merged_df.rename(columns={'sentence_x': 'English_Sentence', 'sentence_y': 'Hindi_Sentence'}, inplace=True)

# Save the merged DataFrame to a new CSV file
merged_csv_filename = 'merged_flores200_data.csv'
merged_df.to_csv(merged_csv_filename, index=False)
print(f"Data has been saved to {merged_csv_filename}")

import pandas as pd

# Load the merged CSV file into a new DataFrame
merged_df = pd.read_csv('merged_flores200_data.csv')

# Print the first 5 rows of the merged DataFrame for verification
print(merged_df.head())


Data has been saved to merged_flores200_data.csv
   id                                              URL_x    domain     topic  \
0   1  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
1   2  https://en.wikinews.org/wiki/Scientists_say_ne...  wikinews    health   
2   3  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
3   4  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   
4   5  https://en.wikinews.org/wiki/Fighter_jet_crash...  wikinews  accident   

   has_image_x  has_hyperlink_x  \
0            0                0   
1            0                0   
2            0                0   
3            0                0   
4            0                0   

                                    English_Sentence  \
0  On Monday, scientists from the Stanford Univer...   
1  Lead researchers say this may bring early dete...   
2  The JAS 39C Gripen crashed onto a runway at ar...   
3  The pilot was identified as Squa

In [6]:
# Print the columns of the merged DataFrame
print(merged_df.columns)

Index(['id', 'URL_x', 'domain', 'topic', 'has_image_x', 'has_hyperlink_x',
       'English_Sentence', 'URL_y', 'has_image_y', 'has_hyperlink_y',
       'Hindi_Sentence'],
      dtype='object')


In [7]:
# Drop unwanted columns
columns_to_drop = ['URL_x', 'has_image_x', 'has_hyperlink_x', 'URL_y', 'has_image_y', 'has_hyperlink_y']
merged_df = merged_df.drop(columns=columns_to_drop)

# Print the updated columns of the merged DataFrame
print(merged_df.columns)


Index(['id', 'domain', 'topic', 'English_Sentence', 'Hindi_Sentence'], dtype='object')


Downloading Helsinki-NLP/opus-mt-en-hi model

In [8]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM
import pickle
translator_opus = pipeline("translation_en_to_hi", model="Helsinki-NLP/opus-mt-en-hi", device=-1)
# Input an English sentence
english_sentence = input("Enter an English sentence: ")
# Translate the input sentence into Hindi using different models
hindi_translation_opus = translator_opus(english_sentence, max_length=50)[0]['translation_text']
# Print the translations
print("Translation (opus):", hindi_translation_opus)

config.json:   0%|          | 0.00/1.39k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/306M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/44.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/812k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/1.07M [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.10M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/models/marian/tokenization_marian.py:197: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


Enter an English sentence: I have a pen in my left hand
Translation (opus): मैं अपने बाएं हाथ में एक कलम है


Downloading facebook/m2m100_418M model

In [9]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM
import pickle
translator_m2m100 = pipeline("translation_en_to_hi", model="facebook/m2m100_418M", device=-1)
# Input an English sentence
english_sentence = input("Enter an English sentence: ")
hindi_translation_m2m100 = translator_m2m100(english_sentence, max_length=50)[0]['translation_text']
print("Translation (m2m100):", hindi_translation_m2m100)


config.json:   0%|          | 0.00/908 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.94G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/233 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/272 [00:01<?, ?B/s]

vocab.json:   0%|          | 0.00/3.71M [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.14k [00:00<?, ?B/s]

Enter an English sentence: I have a pen in my left hand
Translation (m2m100): मेरे बाएं हाथ में एक पेंसिल है


Downloading facebook/mbart-large-50-many-to-many-m model

In [10]:
from transformers import pipeline, AutoTokenizer, AutoModelForSeq2SeqLM
import pickle
translator_mbart = pipeline("translation", model="facebook/mbart-large-50-many-to-many-mmt", device=-1, src_lang="en_XX", tgt_lang="hi_IN")
# Input an English sentence
english_sentence = input("Enter an English sentence: ")
hindi_translation_mbart = translator_mbart(english_sentence, max_length=50)[0]['translation_text']
print("Translation (mbart):", hindi_translation_mbart)


config.json:   0%|          | 0.00/1.43k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.44G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/261 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/529 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/649 [00:00<?, ?B/s]

Enter an English sentence: I have a pen in my left hand
Translation (mbart): मेरे बाएँ हाथ में एक पेन है


In [11]:
# Importing necessary libraries
from transformers import pipeline

# Select only the first 5 rows
merged_df_subset = merged_df.head(5)

# Translate English sentences to Hindi using different models
merged_df_subset['Hindi_Translation_opus'] = merged_df_subset['English_Sentence'].apply(translator_opus)
merged_df_subset['Hindi_Translation_m2m100'] = merged_df_subset['English_Sentence'].apply(translator_m2m100)
merged_df_subset['Hindi_Translation_mbart'] = merged_df_subset['English_Sentence'].apply(translator_mbart)

# Print or further process the translated DataFrame
print(merged_df_subset[['English_Sentence', 'Hindi_Translation_opus', 'Hindi_Translation_m2m100', 'Hindi_Translation_mbart']])


<ipython-input-11-74ae50d00475>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df_subset['Hindi_Translation_opus'] = merged_df_subset['English_Sentence'].apply(translator_opus)
<ipython-input-11-74ae50d00475>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df_subset['Hindi_Translation_m2m100'] = merged_df_subset['English_Sentence'].apply(translator_m2m100)


                                    English_Sentence  \
0  On Monday, scientists from the Stanford Univer...   
1  Lead researchers say this may bring early dete...   
2  The JAS 39C Gripen crashed onto a runway at ar...   
3  The pilot was identified as Squadron Leader Di...   
4  Local media reports an airport fire vehicle ro...   

                              Hindi_Translation_opus  \
0  [{'translation_text': 'यह एक छोटी - सी किस्म क...   
1  [{'translation_text': 'सबसे आगे के खोजकर्ताओं ...   
2  [{'translation_text': 'JS 39C ने 930 के आस - प...   
3  [{'translation_text': 'इस विमान के पायलट को स्...   
4  [{'translation_text': 'स्थानीय मीडिया रिपोर्ट ...   

                            Hindi_Translation_m2m100  \
0  [{'translation_text': 'सोमवार को, स्टैनफोर्ड य...   
1  [{'translation_text': 'प्रमुख शोधकर्ताओं का कह...   
2  [{'translation_text': 'JAS 39C Gripen 9:30 बजे...   
3  [{'translation_text': 'पायलट को स्क्वाड्रन लीड...   
4  [{'translation_text': 'स्थानीय मीडिया रिपोर

<ipython-input-11-74ae50d00475>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df_subset['Hindi_Translation_mbart'] = merged_df_subset['English_Sentence'].apply(translator_mbart)


In [12]:
# Print each column of the first row in a separate row
first_row = merged_df_subset.iloc[0]
for column, value in first_row.items():
    print(f"{column}: {value}")

id: 1
domain: wikinews
topic: health
English_Sentence: On Monday, scientists from the Stanford University School of Medicine announced the invention of a new diagnostic tool that can sort cells by type: a tiny printable chip that can be manufactured using standard inkjet printers for possibly about one U.S. cent each.
Hindi_Sentence: सोमवार को, स्टैनफ़ोर्ड यूनिवर्सिटी स्कूल ऑफ़ मेडिसिन के वैज्ञानिकों ने एक नए डायग्नोस्टिक उपकरण के आविष्कार की घोषणा की जो कोशिकाओं को उनके प्रकार के आधार पर छाँट सकता है: एक छोटी प्रिंट करने योग्य चिप जिसे स्टैण्डर्ड इंकजेट प्रिंटर का उपयोग करके लगभग एक अमेरिकी सेंट के लिए निर्मित किया जा सकता है.
Hindi_Translation_opus: [{'translation_text': 'यह एक छोटी - सी किस्म की छपाई मशीन है, जो अमरीका के हिसाब से अमरीका के हिसाब से अमरीका के हिसाब से एक - से - ज़्यादा लोगों के लिए एक छोटा - सा छपाई मशीन तैयार की जा सकती है ।'}]
Hindi_Translation_m2m100: [{'translation_text': 'सोमवार को, स्टैनफोर्ड यूनिवर्सिटी स्कूल ऑफ मेडिसिन के वैज्ञानिकों ने एक नया निदान उपकरण का

Example 1: New Diagnostic Tool Announcement

Opus-MT Translation:
Translation appears to be nonsensical and does not convey the original meaning accurately.

m2m100 Translation:
Translation is coherent and captures the essence of the original English sentence.

mbart Translation:
Translation is mostly accurate, but there is a mention of "Pazartesi," which is likely a misinterpretation.

In [13]:
# Print each column of the first row in a separate row
first_row = merged_df_subset.iloc[2]
for column, value in first_row.items():
    print(f"{column}: {value}")

id: 3
domain: wikinews
topic: accident
English_Sentence: The JAS 39C Gripen crashed onto a runway at around 9:30 am local time (0230 UTC) and exploded, closing the airport to commercial flights.
Hindi_Sentence: स्थानीय समय (0230 UTC) के मुताबिक करीब 9:30 बजे JAS 39C Gripen रनवे पर धमाके के साथ दुर्घटनाग्रस्त हो गया, जिसकी वजह से वाणिज्यिक उड़ानों के लिए हवाई अड्डे को बंद कर दिया गया.
Hindi_Translation_opus: [{'translation_text': 'JS 39C ने 930 के आस - पास एक रनवे पर धावा बोला, जो स्थानीय समय पर होता है (0230 के दशक) और विस्फोटों से भरा हुआ था ।'}]
Hindi_Translation_m2m100: [{'translation_text': 'JAS 39C Gripen 9:30 बजे स्थानीय समय (0230 यूटीसी) के आसपास एक रेलवे पर टूट गया और विस्फोट हुआ, वाणिज्यिक उड़ानों के लिए हवाई अड्डे को बंद कर दिया।'}]
Hindi_Translation_mbart: [{'translation_text': 'जैस 39सी ग्रिपेन ने 9:30 बजे स्थानीय समय (0230 UTC) पर एक रनवे पर crashed किया और विस्फोट कर हवाई अड्डे को वाणिज्यिक उड़ानों के लिए बंद कर दिया।'}]


Example 2: JAS 39C Gripen Crash

Opus-MT Translation:
Translation is somewhat understandable but lacks precision and fluency.

m2m100 Translation:
Translation is accurate and effectively communicates the events described in the English sentence.

mbart Translation:
Translation is accurate and provides a clear representation of the original sentence.

BLEU Score

In [14]:
from sacrebleu import corpus_bleu

# Assuming you have reference translations in Hindi for your dataset
reference_translations = merged_df_subset['Hindi_Sentence'].tolist()

# Calculate BLEU score for each model
translations_opus = merged_df_subset['Hindi_Translation_opus'].tolist()
translations_m2m100 = merged_df_subset['Hindi_Translation_m2m100'].tolist()
translations_mbart = merged_df_subset['Hindi_Translation_mbart'].tolist()

# Convert each element to string
translations_opus = [str(x) for x in translations_opus]
translations_m2m100 = [str(x) for x in translations_m2m100]
translations_mbart = [str(x) for x in translations_mbart]

# Calculate BLEU score for each translation
bleu_opus = corpus_bleu(translations_opus, [reference_translations])
bleu_m2m100 = corpus_bleu(translations_m2m100, [reference_translations])
bleu_mbart = corpus_bleu(translations_mbart, [reference_translations])

# Print the BLEU scores
print(f'BLEU Score for Opus-MT: {bleu_opus.score}')
print(f'BLEU Score for m2m100: {bleu_m2m100.score}')
print(f'BLEU Score for mbart: {bleu_mbart.score}')


BLEU Score for Opus-MT: 2.256547238077094
BLEU Score for m2m100: 21.820076466821448
BLEU Score for mbart: 20.33483667355113


Opus-MT:-  2.26

A BLEU score around 2 is relatively low. Opus-MT might be struggling with the translation task, possibly due to the complexity of the sentences, domain-specific language, or other issues.

m2m100:-  21.82

A BLEU score of 21.82 is quite high, suggesting good translation quality. m2m100 is performing well on your dataset.

mbart:-  20.33

A BLEU score of 20.33 is also relatively high, indicating good translation quality. mbart seems to be performing well.